tests/karate/src/test/java/common/net/MultiMsgWebSocketClient.java

```diff
@@ -44,24 +49,45 @@ public MessageBuffer getBuffer() {
     return queue;
   }
 
+
   public void publish(String data, String channel){
-    JsonConverter jsonConverter = new JsonConverter();
     Random random = new Random();
     int id = random.nextInt();
+    idAssociatedWithSentMessages.put(data, id);
     Json request =  jsonConverter.publishМessageFromData(data, id, channel);
     this.send(request.toString());
   }
 
-  public String getBackendResponseWithBroadcast(){
-    String answer1 = getBuffer().takeTimeout(5000);
-    String answer2 = getBuffer().takeTimeout(5000);
-    String result = answer1.contains("result") ? answer1 : answer2;
-    return result;
+  public void changeSenderToBeNonAttendee(){
+    jsonConverter.setSenderSk(nonAttendeeSkHex);
+    jsonConverter.setSenderPk(nonAttendeePk);
   }
 
-  public String getBackendResponseWithoutBroadcast(){
-    String result = getBuffer().takeTimeout(5000);
-    return result;
+  public String getBackendResponse(String data){
+    assert idAssociatedWithSentMessages.containsKey(data);
+    int idData = idAssociatedWithSentMessages.get(data);
+    if (idAssociatedWithAnswers.containsKey(idData)){
+      String answer = idAssociatedWithAnswers.get(idData);
+      idAssociatedWithAnswers.remove(idData);
+      idAssociatedWithSentMessages.remove(data);
+      return answer;
+    }
+    String answer = getBuffer().takeTimeout(5000);
+    while(answer != null){
+      if(answer.contains("result") || answer.contains("error")){
+        Json resultJson = Json.of(answer);
+        int idResult = resultJson.get("id");
+        if (idData == idResult){
+          idAssociatedWithSentMessages.remove(data);
+          return answer;
+        }else{
+          idAssociatedWithAnswers.put(idResult, answer);
+        }
+      }
+      answer = getBuffer().takeTimeout(5000);
+    }
```

Comment:
There is a function defined in the buffer that lets you retrieve a message based on a specific filter.

This might be useful here.

(Some filters are already defined in a class)

Reply:
I have an upcoming PR for election self contained tests, I will address this refactoring there, thanks for pointing it out.